### ISSUE TITLE
Incomplete Cover art chunking over UDP

### ISSUE THREAD

**Issue by OneBobone** [mikebrady/shairport-sync#825]
Mostly,  but not always, I only get the first chunk and thereby incomplete images. I have tried with different packet sizes, but have not noticed any difference.
Also, I almost never get the “picture stream has ended code”: “ssncpcen”.

Any recommendations?

**Comment by mikebrady** [comment]
Thanks for the post. Can you get it to misbehave reliably?

**Comment by OneBobone** [comment]
Thank you for responding Mike. 
I would say that it is quite “reliable”, unfortunately.

**Comment by mikebrady** [comment]
Okay, I don't use this facility, so I have added a debug message when each chunk is transmitted and a delay of 100 milliseconds after each chunk. It's in a new branch called `troubleshooting/multicast`. Would you be kind enough to try it out please, and perhaps post some logs?

**Comment by OneBobone** [comment]
Thank you! Let me try that on Tuesday/Wednesday.

**Comment by OneBobone** [comment]
Noticed this issue when installing Cygwin according to your new write-up:
Problem 1/1
package openssl-devel-1.1.1b-1 requires libssl-devel, but none of the providers can be installed
Solution 1/1 (default)
  - do not ask to install openssl-devel-1.1.1b-1

**Comment by mikebrady** [comment]
Thanks. I'll check it out.

The "postinstall script problems" seem to be irrelevant.

**Comment by OneBobone** [comment]
Is this not the right syntax?
git clone https://github.com/mikebrady/shairport-sync.git@troubleshooting/multicast

**Comment by mikebrady** [comment]
What I would do is clone it in the normal way, then `cd` into the new directory and then:
```
$ git checkout troubleshooting/multicast
```

**Comment by OneBobone** [comment]
Instead of "development" additionally?

**Comment by mikebrady** [comment]
Yes, checkout `troubleshooting/multicast` Instead of `development`.

You should be following the steps in the [`development` version of the CYGWIN guide](https://github.com/mikebrady/shairport-sync/blob/development/CYGWIN.md).

**Comment by OneBobone** [comment]
Fair enough, but the CYGWIN guide. for /multicast and development are identical?

**Comment by mikebrady** [comment]
Yes indeed. The multicast branch is a small variation of the development version.

**Comment by mikebrady** [comment]
Hmm. Just thinking of a checklist, can you say where the installation fails?
1. Is Bonjour Service working?
2. Is the CYGWIN D-Bus system service active? 
3. Is the CYGWIN Avahi service active?
4. Did you get `libconfig` successfully installed?

Or, if you have an error message, perhaps you'd be kind enough to post it.

**Comment by OneBobone** [comment]
1-4: yes

I posted the error message above. Is there something else that I can provide?

**Comment by mikebrady** [comment]
Thanks. And did you do the Cygwin setup:
```
C:\Users\mike> Downloads\setup-x86_64.exe -P cygrunsrv,dbus,avahi,avahi-tools,gnome-keyring,libavahi-client-devel,^
libglib2.0-devel,openssl,pkg-config,autoconf,automake,clang,libdaemon-devel,popt-devel,^
make,libao-devel,openssl-devel,libtool,git,wget,flex,bison
```
beforehand?

**Comment by OneBobone** [comment]
Yes, all in one row w/o the ^.
That produced the first posted error.

**Comment by mikebrady** [comment]
Okay. If you copy and paste that:
```
Downloads\setup-x86_64.exe -P cygrunsrv,dbus,avahi,avahi-tools,gnome-keyring,libavahi-client-devel,^
libglib2.0-devel,openssl,pkg-config,autoconf,automake,clang,libdaemon-devel,popt-devel,^
make,libao-devel,openssl-devel,libtool,git,wget,flex,bison
```
into a Windows Command Prompt, including the `^` symbols, it should accept the entry as one line.

That problem:
```
Problem 1/1
package openssl-devel-1.1.1b-1 requires libssl-devel, but none of the providers can be installed
Solution 1/1 (default)

do not ask to install openssl-devel-1.1.1b-1
```
seems to be a Cygwin bug -- I get it too. That may be causing the remaining problems you are experiencing, since it looks like it would not install `libssl` and `libcrypto`, as they are part of the `openssl-devel` package. I suggest you try it again like this on a Windows Command Prompt:

```
Downloads\setup-x86_64.exe -P openssl-devel
```

**Comment by OneBobone** [comment]
`Downloads\setup-x86_64.exe -P openssl-devel` produced the same error.
By downgrading, I manged to get it to work.
![SSL](https://user-images.githubusercontent.com/37246961/55515072-ee3c6100-5669-11e9-914c-eb97a86b862e.jpg)

With the develop branch I got the same chunking error.
I did notice something new: coverart is saved in a cacher folder that was not present before. The images are complete.

Now I will install the other branch and collect logs.

**Comment by OneBobone** [comment]
[shairport-sync.log](https://github.com/mikebrady/shairport-sync/files/3041012/shairport-sync.log)

Not sure how you read the log (I can't see any error), but I fastforwarded around 10 Spotify songs (~ 20 images: 1 small "Spotify std playlist image" and then 1 larger track image.)
The last one appears almost complete on the receiving side. 
Otherwise a good mix of 50% images and 90%-100% images (hard to tell between 90-100%).

If I could receive the `ssnccpen ` it would be nice too.

**Comment by mikebrady** [comment]
Thanks. The log shows that every single chunk is being sent. For example, the very first one:
```
metadata_process: sending chunk 0 of 6.
Connection 1: RTSP Response:
  Type: "CSeq", content: "9"
  Type: "Server", content: "AirTunes/105.1"
Connection 1: Received an RTSP Packet of type "SET_PARAMETER":
  Type: "RTP-Info", content: "rtptime=1170569481"
  Type: "Content-Length", content: "43"
  Type: "Content-Type", content: "text/parameters"
  Type: "CSeq", content: "10"
  Type: "DACP-ID", content: "703406F155B676BE"
  Type: "Active-Remote", content: "779568459"
  Type: "User-Agent", content: "AirPlay/373.9.1"
Connection 1: RTSP Response:
  Type: "CSeq", content: "10"
  Type: "Server", content: "AirTunes/105.1"
Connection 1: Received an RTSP Packet of type "SET_PARAMETER":
  Type: "RTP-Info", content: "rtptime=1170569481"
  Type: "Content-Length", content: "189"
  Type: "Content-Type", content: "application/x-dmap-tagged"
  Type: "CSeq", content: "11"
  Type: "DACP-ID", content: "703406F155B676BE"
  Type: "Active-Remote", content: "779568459"
  Type: "User-Agent", content: "AirPlay/373.9.1"
Connection 1: RTSP Response:
  Type: "CSeq", content: "11"
  Type: "Server", content: "AirTunes/105.1"
metadata_process: sending chunk 1 of 6.
metadata_process: sending chunk 2 of 6.
metadata_process: sending chunk 3 of 6.
metadata_process: sending chunk 4 of 6.
Connection 1: Received an RTSP Packet of type "SET_PARAMETER":
  Type: "RTP-Info", content: "rtptime=1170569481"
  Type: "Content-Length", content: "89346"
  Type: "Content-Type", content: "image/jpeg"
  Type: "CSeq", content: "12"
  Type: "DACP-ID", content: "703406F155B676BE"
  Type: "Active-Remote", content: "779568459"
  Type: "User-Agent", content: "AirPlay/373.9.1"
Connection 1: RTSP Response:
  Type: "CSeq", content: "12"
  Type: "Server", content: "AirTunes/105.1"
Connection 1: Received an RTSP Packet of type "SET_PARAMETER":
  Type: "RTP-Info", content: "rtptime=1170586025"
  Type: "Content-Length", content: "43"
  Type: "Content-Type", content: "text/parameters"
  Type: "CSeq", content: "13"
  Type: "DACP-ID", content: "703406F155B676BE"
  Type: "Active-Remote", content: "779568459"
  Type: "User-Agent", content: "AirPlay/373.9.1"
Connection 1: RTSP Response:
  Type: "CSeq", content: "13"
  Type: "Server", content: "AirTunes/105.1"
Connection 1: Received an RTSP Packet of type "SET_PARAMETER":
  Type: "RTP-Info", content: "rtptime=1170586025"
  Type: "Content-Length", content: "189"
  Type: "Content-Type", content: "application/x-dmap-tagged"
  Type: "CSeq", content: "14"
  Type: "DACP-ID", content: "703406F155B676BE"
  Type: "Active-Remote", content: "779568459"
  Type: "User-Agent", content: "AirPlay/373.9.1"
Connection 1: RTSP Response:
  Type: "CSeq", content: "14"
  Type: "Server", content: "AirTunes/105.1"
metadata_process: sending chunk 5 of 6.
MH Picture received, length 89346 bytes.
```
You can see all six chunks are sent (the index number is always 1 less than it should be in the log).

I've updated the branch to fix that cosmetic issue and add more diagnostic output. My guess is that UDP packets are being lost on your network -- remember UDP does not guarantee delivery.

The cached files are for the D-Bus interface and are not part of the metadata stream supplyied by the source.

**Comment by OneBobone** [comment]
Ok, I can see. I will give it another try tonight. A few thoughts:
- no other udp packets are lost other than the last image chunk. Too much of a coincidence?
- where is the ssncpcen package? Is it being concatenated with the last chunk?
- the cached file could be of interest as a back-up solution...

**Comment by mikebrady** [comment]
The changes I made should show that the last chunk and the `pcen` item are being sent, but of course I'll keep looking for any issues. What is receiving the metadata?

**Comment by OneBobone** [comment]
A .Net standard udp receiver running on the same machine.
Will try later this evening.

**Comment by OneBobone** [comment]
[shairport-sync.log](https://github.com/mikebrady/shairport-sync/files/3047089/shairport-sync.log)
Can't see any errors.

Want to bring your attention to my log below. 
> apr-05 10:49:31	 	ssncpcen	ssncpcen3291532220
> apr-05 10:49:30	 	ssncpcst	ssncpcst3291532220
> apr-05 10:49:30	 	Playing: 1671 Mike Perry Hands
> apr-05 10:49:20	 	ssncpcen	ssncpcen3291045756
> apr-05 10:49:19	 	ssncpcst	ssncpcst3291045756
> apr-05 10:49:19	 	ssncpcst	ssncpcst3291022172
> apr-05 10:49:18	 	Playing: 2454 The Script Superheroes
> apr-05 10:49:06	 	ssncpcen	ssncpcen3290416028
> apr-05 10:49:05	 	ssncpcst	ssncpcst3290416028
> apr-05 10:49:04	 	Playing: 1826 Ellie Goulding Close To Me (feat. Swae Lee)
> apr-05 10:48:55	 	ssncpcen	ssncpcen3289984828
> apr-05 10:48:55	 	ssncpcst	ssncpcst3289984828
> apr-05 10:48:55	 	ssncpcst	ssncpcst3289960540
> apr-05 10:48:54	 	Playing: 2070 Zedd Happy Now


Could it be that for unchunked (small) pictures, `ssncpcen` and sometimes goes `ssncpcst `missing? Typically for the Spotify playlist image.

**Comment by mikebrady** [comment]
Ha, thanks for that. I just pushed a slight update with a bit more error logging on sending the UDP packets -- it might be interesting, I just don't know.

I'll have a look at the issue you raise.

**Comment by mikebrady** [comment]
That looks like an old log? It doesn't seem to include any of the extra log information included in yesterday's update.

**Comment by OneBobone** [comment]
Will reinstall. (Sorry if I mixed up the logs, but they did not look so different to me either.
Could ssncpcen be received before the last chunk (chunking sub)?

**Comment by mikebrady** [comment]
Thanks.

Right off the top of my head, we know that UDP packets can be delivered out-of-order, so there is a way this could happen.

As far as Shairport Sync itself is concerned, one thread is concerned with receiving the metadata and it hands off the processing of it, in the order of reception, to a single other thread. The metadata processing thread processes everything in order, so I can't see where an out-of-order effect could take place within Shairport Sync. But once the UDP packets are emitted, anything could happen.

**Comment by OneBobone** [comment]
> 
> apr-05 11:35:24	 	SPS	ssncpcen2472336411
> apr-05 11:35:24	 	SPS	ssncpcst2472336411
> apr-05 11:35:23	 	SPS	ssncpcst2472295227
> apr-05 11:35:22	 	Nuvo Airplay	Playing: 1983 G.R.L. Ugly Heart
> apr-05 11:31:46	 	SPS	ssncpcen2462694427
> apr-05 11:31:45	 	SPS	ssncpcst2462694427
> apr-05 11:31:44	 	SPS	ssncpcst2462650075
> apr-05 11:31:43	 	Nuvo Airplay	Playing: 2182 James Arthur Rewrite The Stars (with James Arthur & Anne-Marie)
> apr-05 11:28:16	 	SPS	ssncpcen2453488571
> apr-05 11:28:16	 	SPS	ssncpcst2453488571
> apr-05 11:28:16	 	SPS	ssncpcst2453466043
> apr-05 11:28:15	 	Nuvo Airplay	Playing: 2077 Jason Derulo Want to Want Me
> apr-05 11:27:41	 	SPS	ssncpcst2451813403
> apr-05 11:27:41	 	Nuvo Airplay	Playing: 10720 1990 WALK THE MOON Shut Up and Dance
> apr-05 11:27:40	 	Nuvo Airplay	Paused: 10720 1990 WALK THE MOON Shut Up and Dance
> apr-05 11:27:40	 	SPS	ssncpcen2451800893
> apr-05 11:27:40	 	SPS	ssncpcst2451800893
> apr-05 11:27:39	 	Nuvo Airplay	Playing: 10710 1990 WALK THE MOON Shut Up and Dance
> apr-05 11:27:39	 	Nuvo Airplay	Paused: 10710 1990 WALK THE MOON Shut Up and Dance
> apr-05 11:27:39	 	Nuvo Airplay	Bob's iPad

[shairport-sync.log](https://github.com/mikebrady/shairport-sync/files/3047285/shairport-sync.log)

**Comment by OneBobone** [comment]
Dear Mike, I am not sure I want to waste your precious time on this.
The fact that the Cover art is cached is a good back-up.
If I may, getting dbus working would be very valuable.
:-)

**Comment by mikebrady** [comment]
Thanks. Apologies for the bug-laden previous update. I've fixed it.

The D-Bus situation is frustrating alright, but will take some time to resolve, as, AFAICT, it's either a CYGWIN or a D-Bus issue. I need to prepare a simple exemplar of the problem and ask the experts...

**Comment by OneBobone** [comment]
Pls Excuse, which bug did you fix so I can test.

**Comment by OneBobone** [comment]
2 questions: 
- The Linux Pipe becomes a windows link. Not sure how to get the metadata to a file? I tried replacing/adding to the pipe "metadata.txt"
- does the pipe and UPD output run in parallel or are they exclusive?

**Comment by mikebrady** [comment]
There were two bugs -- one was incorrectly reporting an error every time a piece of metadata was sent out on UDP. The second was mis-formatting the code of a piece of metadata.


* The metadata pipe is, by default, at `/tmp/shairport-sync-metadata`. In Linux, you could direct it to a file like this:
```
$ cat /tmp/shairport-sync-metadata > file.txt
```
* If metadata is enabled, both the pipe and UDP are simultaneously available.

**Comment by OneBobone** [comment]
Thank you.
To summarize what we have learned:
1) Installing on Cygwin is currently not 100% straight forward as the SSL versions are not mutually compatible.
--> the solution is to select the next latest version of these
2) Reception of `ssncpcen` is not 100% consistent. 
- it is often (always) lacking for small images that do not need chunking. Typically the Spotify playlist logo that is sent between songs.
- when the "system" is stressed, e.g. when forcing the next song with a "Next" command, the last `ssncpcen` can also be lacking. Unclear if the whole chunk is malformed/malreceived as well.

**Comment by mikebrady** [comment]
Yes, agreed. I think item 1 is a transient bug in Cygwin. Regarding item 2, it seems as if Shairport Sync is outputting the stuff to UDP but it is disappearing.
Item 3 is a bit surprising. Does it `$ cat` to the console? (That is, does:  `$ cat /tmp/shairport-sync-metadata` work?)

**Comment by mikebrady** [comment]
Closing this interesting thread – I think we've moved on. Please feel free to open another issue if necessary.
